editor/plugins/gizmos/collision_shape_3d_gizmo_plugin.cpp

```diff
@@ -424,64 +424,95 @@ void CollisionShape3DGizmoPlugin::redraw(EditorNode3DGizmo *p_gizmo) {
-			points.push_back(Vector3(b.x, 0, b.y) - d);
-
-			if (i % 90 == 0) {
-				points.push_back(Vector3(a.x, 0, a.y) + d);
-				points.push_back(Vector3(a.x, 0, a.y) - d);
-			}
-
-			Vector3 dud = i < 180 ? d : -d;
-
-			points.push_back(Vector3(0, a.x, a.y) + dud);
-			points.push_back(Vector3(0, b.x, b.y) + dud);
-			points.push_back(Vector3(a.y, a.x, 0) + dud);
-			points.push_back(Vector3(b.y, b.x, 0) + dud);
+		Vector<Vector3> points;
+		// 4 vertical lines and 4 full circles.
+		points.resize(4 * 2 + 4 * 8 * points_in_octant * 2);
+		Vector3 *points_ptrw = points.ptrw();
+
+		uint32_t index = 0;
+		float y_value = height * 0.5 - radius;
+
+		// Vertical Lines.
+		points_ptrw[index++] = Vector3(0.f, y_value, radius);
+		points_ptrw[index++] = Vector3(0.f, -y_value, radius);
+		points_ptrw[index++] = Vector3(0.f, y_value, -radius);
+		points_ptrw[index++] = Vector3(0.f, -y_value, -radius);
+		points_ptrw[index++] = Vector3(radius, y_value, 0.f);
+		points_ptrw[index++] = Vector3(radius, -y_value, 0.f);
+		points_ptrw[index++] = Vector3(-radius, y_value, 0.f);
+		points_ptrw[index++] = Vector3(-radius, -y_value, 0.f);
+
+#define PUSH_QUARTER(from_x, from_y, to_x, to_y, y)      \
+	points_ptrw[index++] = Vector3(from_x, y, from_y);   \
+	points_ptrw[index++] = Vector3(to_x, y, to_y);       \
+	points_ptrw[index++] = Vector3(from_x, y, -from_y);  \
+	points_ptrw[index++] = Vector3(to_x, y, -to_y);      \
+	points_ptrw[index++] = Vector3(-from_x, y, from_y);  \
+	points_ptrw[index++] = Vector3(-to_x, y, to_y);      \
+	points_ptrw[index++] = Vector3(-from_x, y, -from_y); \
+	points_ptrw[index++] = Vector3(-to_x, y, -to_y);
+
+#define PUSH_QUARTER_XY(from_x, from_y, to_x, to_y, y)       \
+	points_ptrw[index++] = Vector3(from_x, -from_y - y, 0);  \
+	points_ptrw[index++] = Vector3(to_x, -to_y - y, 0);      \
+	points_ptrw[index++] = Vector3(from_x, from_y + y, 0);   \
+	points_ptrw[index++] = Vector3(to_x, to_y + y, 0);       \
+	points_ptrw[index++] = Vector3(-from_x, -from_y - y, 0); \
+	points_ptrw[index++] = Vector3(-to_x, -to_y - y, 0);     \
+	points_ptrw[index++] = Vector3(-from_x, from_y + y, 0);  \
+	points_ptrw[index++] = Vector3(-to_x, to_y + y, 0);
+
+#define PUSH_QUARTER_YZ(from_x, from_y, to_x, to_y, y)       \
+	points_ptrw[index++] = Vector3(0, -from_y - y, from_x);  \
+	points_ptrw[index++] = Vector3(0, -to_y - y, to_x);      \
+	points_ptrw[index++] = Vector3(0, from_y + y, from_x);   \
+	points_ptrw[index++] = Vector3(0, to_y + y, to_x);       \
+	points_ptrw[index++] = Vector3(0, -from_y - y, -from_x); \
+	points_ptrw[index++] = Vector3(0, -to_y - y, -to_x);     \
+	points_ptrw[index++] = Vector3(0, from_y + y, -from_x);  \
+	points_ptrw[index++] = Vector3(0, to_y + y, -to_x);
+
+		float previous_x = radius;
+		float previous_y = 0.f;
+
+		for (uint32_t i = 0; i < points_in_octant; ++i) {
+			r += inc;
+			real_t x = Math::cos(r) * radius;
+			real_t y = Math::sqrt(radius_squared - (x * x));
```

Comment:
This has the same gaps as the cylinder (https://github.com/godotengine/godot/pull/103903#issuecomment-2712086721).

![godot windows editor x86_64_Tfji0gABxd](https://github.com/user-attachments/assets/b8d0ea11-2ea6-4882-af5a-271fcd777f6c)|![godot windows editor x86_64_OL3N3q9Srf](https://github.com/user-attachments/assets/c1706759-3d16-46b3-a614-d67b6ab8e542)
-|-


A fix would be as easy as someting like this:
```suggestion
			real_t y = (i == points_in_octant - 1) ? x : Math::sqrt(radius_squared - (x * x));
```

Reply:
Thanks for spotting this ! I can definitely see the gap on your screenshots, but can't reproduce locally. I tried forcing the final angle to be PI/4, because I guess th error comes in the accumulation in r. If you could check the new version to see if you still see gaps, it would be helpful because I have no way of validating it locally.

Reply:
I've tested the new version, it's the same. But note that forcing the angle to (approx) PI/4 (as it's currently done) doesn't guarantee that `x == y`, these can still differ by some eps:
```cpp
x = Math::cos(MATH_PI / 4) * radius;
y = Math::sqrt((radius * radius) - (x * x));
```

---

However, I've tested my original suggested version as well, which ensures the last points match exactly (if `x == y` then `(x, y) == (y, x)` (I'm skipping the third same-across-whole-circle coord)):
```cpp
real_t x = Math::cos(r) * radius;
real_t y = (i == points_in_octant - 1) ? x : Math::sqrt(radius_squared - (x * x));
```
(hand-drawn, absolute inaccurate image)
![ShareX_quSGsj5eiE](https://github.com/user-attachments/assets/39db6817-cf6b-4735-9342-f705c65376e6)

...and the result was the same, still gaps.

Same for combined version: ensuring PI/4 angle and `x == y`:
```cpp
real_t x = Math::cos((i == points_in_octant - 1) ? octant_angle : r) * radius;
real_t y = (i == points_in_octant - 1) ? x : Math::sqrt(radius_squared - (x * x));
```
Hence I'm concluding the issue is caused by something else, unrelated to this PR. However, I still think it's good to ensure the last points match exactly here, aka that `x == y` for the last iteration.
Leaving it up to you whether you want to ensure the angle as well, or not care at all about this potential slight mismatch (in such case could be always tackled/fixed if reported later). So you can change to whatever desired final version, mark this as resolved, and should be good to merge.

I'll probably try to investigate further what's the actual cause of the gaps, it's baffles me why they're only on the octant edges, even with the lines data having no gaps. :thinking:

Reply:
Makes sense - if the error is big enough to result in a visible gap, then after `y = x` the created bump might also be big enough to be noticable.
My thinking was that forcing exactly equal vertices would make GPU ensure no gaps, but I think I just made this up. :laughing: I don't think that's anyhow guaranteed for `PRIMITIVE_LINES`? Would probably need to use `PRIMITIVE_LINE_STRIP` instead (and not duplicate points in the first place). But, given the same point-pairs are used for collision segments, it does make sense to not generate two separate buffers (`PRIMITIVE_LINE_STRIP` would need only the first point duplicated).

So if ensuring exactly equal vertices doesn't guarantee anything GPU-wise, then it shouldn't be necessary. Just forcing the PI/4 angle should already greatly reduce potential error indeed.

Reply:
Too bad we didn't fix it though. I'm resolving this for now but if you have another idea on how to fix this (or just the real cause), just comment on the PR and I'll look into it.

Reply:
I can reproduce the issue with the gaps. I'll do a quick check in a GPU frame debugger to see if the GPU is getting weird values

Reply:
In Renderdoc I found in one place where a gap forms the vertex positions are:
vertex1: -0.33578, 0.5, 0.37048
vertex2: -0.35355, 0.5, 0.35355

As kleonc guessed above, vertex positions need to be identical between segments in order to be airtight. Any tiny gap can be accumulated by the GPU even if the size of the gap would have been less than should be visible. 

This gizmo is drawing the lines in a bunch of disconnected segments, and since some segments end slightly off from where the next begins, there are gaps forming (most likely due to floating point error in the GPU given the slightly offset vertex positions). 

Overall the issue is pretty subtle and requires some zooming around to reproduce. So I agree that it shouldn't block merging

Reply:
That is a pretty big difference to be a floating point error. Seems unlikely that the computation leads to an error this big. I'll recheck the code one more time.

Reply:
> In Renderdoc I found in one place where a gap forms the vertex positions are:
> vertex1: -0.33578, 0.5, 0.37048
> vertex2: -0.35355, 0.5, 0.35355

> That is a pretty big difference to be a floating point error.

Yeah, these are endpoints of a single line segment, they're not endpoints from neigbor segments ([angle between them on the `y = 0.5` plane is ≈ 2.8128](https://www.wolframalpha.com/input?i=angle+between+%28-0.33578%2C+0.0%2C+0.37048%29+and+%28-0.35355%2C+0.0%2C+0.35355%29) `≈ 2.8125 = 45.0 / 16` (45 = octant angle, 16 = segments per octant)).

I've also checked in RenderDoc, haven't spotted any issue with the vertex data. Now rechecked specific points you've mentioned (they're from the default capsule: radius = 0.5, height = 2.0), note that the should-match last point is at +/- 8 index offset (it's added in a subsequent macro "call"; each macro adds 8 vertices).

Your vertex1/vertex2 are 980/981, and 981 matches 973 just fine:
![xfbqzAKsaf](https://github.com/user-attachments/assets/c926674c-7e4f-4f5a-a72b-acb7440e501d)|![l2HJDAemBF](https://github.com/user-attachments/assets/3adf4566-c80d-409f-b3c7-1168dcd7b4fd)
-|-
![qrenderdoc_LwHBc63Kek](https://github.com/user-attachments/assets/073ae9cd-410c-455a-874b-89991fee9d39)|![LiwCtxBPTQ](https://github.com/user-attachments/assets/4a785606-26a6-4bf4-88dd-2c67bd789a7d)

Went further and exported the binary vertex data from the RenderDoc, all positions besides the intersection points occur exactly 2 times (thus are part of 2 segments as there are no zero length segments). Intersection points occur 4 times each:
```
4   00000000 3f000000 3f000000  [0.0, 0.5, 0.5]
4   00000000 bf000000 3f000000  [0.0, -0.5, 0.5]
4   00000000 3f000000 bf000000  [0.0, 0.5, -0.5]
4   00000000 bf000000 bf000000  [0.0, -0.5, -0.5]
4   3f000000 3f000000 00000000  [0.5, 0.5, 0.0]
4   3f000000 bf000000 00000000  [0.5, -0.5, 0.0]
4   bf000000 3f000000 00000000  [-0.5, 0.5, 0.0]
4   bf000000 bf000000 00000000  [-0.5, -0.5, 0.0]
4   00000000 bf800000 00000000  [0.0, -1.0, 0.0]
4   00000000 3f800000 00000000  [0.0, 1.0, 0.0]
```
All as expected.